### PR TITLE
Remote Facade Pattern

### PR DESCRIPTION
**Pull request by agam1610**
- Added the remote facade design pattern along with tests and readme.
- solves issue #1311 

> For detailed contributing instructions see https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute

### CODE REVIEW

File: remote-facade/src/main/java/com/iluwatar/remotefacade/domain/Domain.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * Class acting as a database to store details in an Arraylist.
+ */
+public class Domain {
+  public static List<Customer> customers = new ArrayList<Customer>();
+  /**
+  * get the details of the customers in the array to ensure correctness.
+  */
+
+  public static ArrayList<String> getCustomers() {
```
Comment (robertvolkmann) [inline remote-facade/src/main/java/com/iluwatar/remotefacade/domain/Domain.java:15]: Why do we need this function?

File: remote-facade/src/main/java/com/iluwatar/remotefacade/domain/Domain.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package com.iluwatar.remotefacade.domain;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Class acting as a database to store details in an Arraylist.
+ */
+public class Domain {
```
Comment (robertvolkmann) [inline remote-facade/src/main/java/com/iluwatar/remotefacade/domain/Domain.java:9]: Maybe rename it to Store or DummyDatabase.

Comment (agam1610) [reply remote-facade/src/main/java/com/iluwatar/remotefacade/domain/Domain.java]: fixed this

File: remote-facade/src/main/java/com/iluwatar/remotefacade/domain/Domain.java (hunk 1 above)
Comment (agam1610) [reply remote-facade/src/main/java/com/iluwatar/remotefacade/domain/Domain.java]: It was a way for me to check all the details were getting added correctly, I forgot to remove it, but it should be fixed now

File: remote-facade/src/test/java/com/iluwatar/remotefacade/test/ClientTest.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import com.iluwatar.remotefacade.domain.Domain;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class ClientTest {
+
+    Customer AMY = new Customer("Amy Adams", "0490490490", "88 Radford Street");
+    Customer JOHN = new Customer("John Adams", "0449944994", "12 Kingsford Street");
```
[the remark this answers is not in the document]
Comment (agam1610) [reply remote-facade/src/test/java/com/iluwatar/remotefacade/test/ClientTest.java]: Fixed this

File: remote-facade/src/main/java/com/iluwatar/remotefacade/domain/Customer.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+/**
+ * {@link Customer } is an object with customer details.
+ */
+@Getter
+@Setter
+@RequiredArgsConstructor
+public class Customer {
```
[the remark this answers is not in the document]
Comment (agam1610) [reply remote-facade/src/main/java/com/iluwatar/remotefacade/domain/Customer.java:7]: Removed all setters andd made the fields final

File: remote-facade/src/main/java/com/iluwatar/remotefacade/dto/Customerdto.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+
+/**
+ * {@link Customerdto } is a data transfer object.
+ *
+ * <p>Dto will not have any business logic in it.
+ */
+@Getter
+@Setter
```
[the remark this answers is not in the document]
Comment (agam1610) [reply remote-facade/src/main/java/com/iluwatar/remotefacade/dto/Customerdto.java]: Removed getters and setters

**Comment by agam1610** [comment]
Hi, I have followed the suggestions and made the changes accordingly.
Thank You

File: remote-facade/README.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
+---
+layout: pattern
+title: Remote Facade Pattern
+folder: remote-facade
+permalink: /patterns/remote-facade/
+categories: Architectural
+language: en
+tags:
+- Performance
+---
```
Comment (iluwatar) [inline remote-facade/README.md:10]: Please check the yaml frontmatter requirements at https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute

File: remote-facade/README.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+> We need to fetch information about customers from a remote database. Instead of making the
+> calls one at a time for certain attributes, we use DTOs to transfer all the relevant attributes in a single call.
+
+In plain words
+
+> Using a DTO relevant information can be fetched with a single call by creating a facade.
+
+Wikipedia says
+
+> Has no information about this pattern.
```
Comment (iluwatar) [inline remote-facade/README.md:29]: Lets leave out in this case

File: remote-facade/pom.xml [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>5.7.1</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (iluwatar) [inline remote-facade/pom.xml:50]: I think this is included in junit-jupiter-engine dependency and can be left out. See https://junit.org/junit5/docs/current/user-guide/#running-tests-build-maven-engines-configure

File: remote-facade/pom.xml [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                </configuration>
+                <executions>
+                    <execution>
+                        <phase>process-classes</phase>
+                        <goals>
+                            <goal>map</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
```
Comment (iluwatar) [inline remote-facade/pom.xml:91]: This doesn't need to be configured here. It's present in the parent pom.xml

File: remote-facade/src/main/java/com/iluwatar/remotefacade/Client.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  public static void main(String[] args) {
+    var amy = new Customer("Amy Adams", " 0490490490", " 88 Radford Street");
+    var john = new Customer("John Adams", " 0449944994", " 12 Kingsford Street");
+
+    RemoteFacade.makeClient(Customerdtoassembler.makeCustomerdto(amy));
+    RemoteFacade.makeClient(Customerdtoassembler.makeCustomerdto(john));
+
+    LOGGER.info("Customer Details: ");
+    printCustomerDetails(Store.customers);
+  }
```
Comment (iluwatar) [inline remote-facade/src/main/java/com/iluwatar/remotefacade/Client.java:33]: Add comments

File: remote-facade/src/main/java/com/iluwatar/remotefacade/Client.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  private static void printCustomerDetails(List<Customer> list) {
+    list.forEach(customer -> LOGGER.info(String.valueOf(Store.customers)));
+  }
```
Comment (iluwatar) [inline remote-facade/src/main/java/com/iluwatar/remotefacade/Client.java:36]: Not sure if this one-liner needs a separate method?

File: remote-facade/src/main/java/com/iluwatar/remotefacade/assemblersfacades/Customerdtoassembler.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+  * create a dataObject for a customer with their details.
+  *
+  * @param cstmr .
+  */
+  public static Customerdto makeCustomerdto(Customer cstmr) {
+    Customerdto customer = new Customerdto();
+    customer.name = cstmr.getName();
+    customer.phone = cstmr.getPhone();
+    customer.address = cstmr.getAddress();
+    return customer;
```
Comment (iluwatar) [inline remote-facade/src/main/java/com/iluwatar/remotefacade/assemblersfacades/Customerdtoassembler.java:20]: Prefer `var` for local variables

File: remote-facade/src/main/java/com/iluwatar/remotefacade/dto/Customerdto.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
+package com.iluwatar.remotefacade.dto;
+
+/**
+ * {@link Customerdto } is a data transfer object.
+ *
+ * <p>Dto will not have any business logic in it.
+ */
+public class Customerdto {
```
Comment (iluwatar) [inline remote-facade/src/main/java/com/iluwatar/remotefacade/dto/Customerdto.java:8]: Consider renaming to CustomerDTO

File: remote-facade/src/test/java/com/iluwatar/remotefacade/test/ClientTest.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+import com.iluwatar.remotefacade.domain.Store;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class ClientTest {
+
+    @Test
+    public void checkClientAMY(){
```
Comment (iluwatar) [inline remote-facade/src/test/java/com/iluwatar/remotefacade/test/ClientTest.java:15]: Consider renaming to checkClientAmy

File: remote-facade/src/main/java/com/iluwatar/remotefacade/assemblersfacades/Customerdtoassembler.java (hunk 12 above)
Comment (iluwatar) [reply remote-facade/src/main/java/com/iluwatar/remotefacade/assemblersfacades/Customerdtoassembler.java:20]: Check throughout the code

**Review by iluwatar — CHANGES_REQUESTED** [review]

**Comment by iluwatar** [comment]
I appreciate the reviewing help @robertvolkmann, thank you!

**Pull request closed without merging**